mlir/lib/Gradient/Transforms/BufferizationPatterns.cpp

```diff
@@ -32,8 +34,25 @@ class BufferizeAdjointOp : public OpConversionPattern<AdjointOp> {
         SmallVector<Type> resTypes;
         if (failed(getTypeConverter()->convertTypes(op.getResultTypes(), resTypes)))
             return failure();
-        rewriter.replaceOpWithNewOp<AdjointOp>(op, resTypes, op.getCalleeAttr(),
-                                               adaptor.getOperands());
+
+        Location loc = op.getLoc();
+        Value gradSize = op.getGradSize();
+        SmallVector<Value> memrefValues;
+        SmallVector<Value> tensorValues;
+        int i = 0;
+        for (Type resType : resTypes) {
+            MemRefType memrefType = resType.cast<MemRefType>();
+            Value memrefValue = rewriter.create<memref::AllocOp>(loc, memrefType, gradSize);
+            Value tensorValue = rewriter.create<bufferization::ToTensorOp>(
+                loc, op.getResultTypes()[i++], memrefValue);
+            memrefValues.push_back(memrefValue);
+            tensorValues.push_back(tensorValue);
+        }
+
+        op.replaceAllUsesWith(tensorValues);
+        rewriter.create<AdjointOp>(loc, TypeRange{}, op.getCalleeAttr(), adaptor.getArgs(),
+                                   adaptor.getGradSize(), memrefValues);
+        rewriter.eraseOp(op);
```

Comment:
Slightly shorter:
```suggestion
        op.replaceAllUsesWith(tensorValues);
        rewriter.create<AdjointOp>(loc, TypeRange{}, op.getCalleeAttr(), adaptor.getArgs(),
                                   adaptor.getGradSize(), memrefValues);
        rewriter.replaceOp(op, memrefValues);
```

Reply:
[66b8310](https://github.com/PennyLaneAI/catalyst/pull/63/commits/66b831097c5b40dd6f81a894dbe50ac04d8ccf7e)

Reply:
I think the reason your first patch doesn't work is because you use `op.replaceAllUsesWith` rather than the `rewriter.replaceOp` that I suggested in the AdjointOp snippet.
Quickly testing it on `SampleOp` and it seems to work:

```cpp
        Value allocVal = rewriter.create<memref::AllocOp>(loc, resultType);
        rewriter.create<SampleOp>(loc, TypeRange{}, ValueRange{adaptor.getObs(), allocVal},
                                  op->getAttrs());
        rewriter.replaceOp(op, allocVal);
```
```mlir
  func.func @counts(%arg0: !quantum.bit, %arg1: !quantum.bit) -> tensor<2x2xf64> {
    %0 = quantum.compbasis %arg0, %arg1 : !quantum.obs
    %alloc = memref.alloc() : memref<2x2xf64>
    quantum.sample %0 in(%alloc : memref<2x2xf64>) {shots = 2 : i64}
    %1 = bufferization.to_tensor %alloc : memref<2x2xf64>
    return %1 : tensor<2x2xf64>
  }
```

Note that besides the above, in general all IR mutation should go through the rewriter: https://mlir.llvm.org/docs/PatternRewriter/#restrictions

Reply:
Thanks! Will change it for SampleOp as well then.

Reply:
Same for all the others as well 👍

Reply:
Oh my bad, I must have been looking at an outdated version, looks good 💯